Add simple text editor widget

erilyth opened the pull request:
Using the CollabWrapperAPI "https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/272", a simple text editor widget which would provide us a Gtk.Box which we can add to our view in any activity.

samdroid-apps commented on src/sugar3/activity/widgets.py:
```diff
@@ -396,3 +396,31 @@ def __init__(self):
+        self.textview.set_wrap_mode(Gtk.WrapMode.WORD)
+        self.textbuffer = self.textview.get_buffer()
+        self.textbuffer.connect('changed', self.__text_buffer_edited_cb)
+        self.textbuffer.set_text("TESTING")
```
Please make it empty by default.  Also, please deal with if a new buddy joins - you need to transfer the text to them.

samdroid-apps commented on src/sugar3/activity/widgets.py:
```diff
@@ -31,9 +31,9 @@
 
-
-_ = lambda msg: gettext.dgettext('sugar-toolkit-gtk3', msg)
+_ = lambda msg: gettext.dgettext('sugar-toolkit-Gtk3', msg)
```
Why?

samdroid-apps commented on src/sugar3/activity/widgets.py:
```diff
@@ -396,3 +396,31 @@ def __init__(self):
+
+    def __message_cb(self, collab, buddy, message):
+        action = message.get('action')
+	if action == 'entry_changed':
```
Can we scope it?  What if I want to have 2 collab text boxes?

erilyth replied:
You mean multiple text boxes should use the same collab setup? I'm a little confused, could you elaborate a bit, thanks.

samdroid-apps replied:
On Sun, Dec 13, 2015, 10:06 PM Batchu Venkat Vishal <
notifications@github.com> wrote:

In src
<https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#discussion_r47443076>
/
<https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#discussion_r47443076>
sugar3
<https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#discussion_r47443076>
/activity/
<https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#discussion_r47443076>
widgets.py
<https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#discussion_r47443076>
:

> + self._collab.connect('message', self.__message_cb) > +
self._collab.setup() > + self.textview = Gtk.TextView() > +
self.textview.set_editable(True) > + self.textview.set_cursor_visible(True)
> + self.textview.set_wrap_mode(Gtk.WrapMode.WORD) > + self.textbuffer =
self.textview.get_buffer() > + self.textbuffer.connect('changed',
self.__text_buffer_edited_cb) > + self.textbuffer.set_text("TESTING") > +
self.textview.show() > + self.pack_start(self.textview, expand=True,
fill=True, padding=0) > + self.show() > + > + def __message_cb(self,
collab, buddy, message): > + action = message.get('action') > + if action
== 'entry_changed':

You mean multiple text boxes should use the same collab setup? I'm a little
confused, could you elaborate a bit, thanks.

Yes.  So add another argument to the parameter that takes an ID for the
text box.  Then prefix all actions with that ID - so if I have 2 IDs they
don't sync.

—
Reply to this email directly or view it on GitHub
<https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274/files#r47443076>.<img
src="
https://ci6.googleusercontent.com/proxy/_aBzafILOCID-2ffcsDd82B8-k1Bvm5np_Fwcwa4UkKLA_k0WScmjkj5NJF-pmu3XuF2g0sBjCPUgfhBHHh8GfY4tJQDmc7UmDfqT-5EnVdjboGGZy4uexfzAySBFfz710aPCyO5iEPS-Rw7niDf5YYuWq4YoQ=s0-d-e1-ft#https://github.com/notifications/beacon/AFvjmhiXkyJsYxyQo8WqprCVhhwxfoGoks5pPUitgaJpZM4G0RrG.gif
">

erilyth replied:
Should I randomly generate the ID? Or is there a better way to assign ID's?

samdroid-apps replied:
I think you should let the activity assign the id.

On Mon, Dec 14, 2015, 4:09 PM Batchu Venkat Vishal <notifications@github.com>
wrote:

> In src/sugar3/activity/widgets.py
> <https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#discussion_r47461956>
> :
>
> > +        self._collab.connect('message', self.__message_cb)
> > +        self._collab.setup()
> > +        self.textview = Gtk.TextView()
> > +        self.textview.set_editable(True)
> > +        self.textview.set_cursor_visible(True)
> > +        self.textview.set_wrap_mode(
>
> Gtk.WrapMode.WORD)
> > +        self.textbuffer = self.textview.get_buffer()
> > +        self.textbuffer.connect('changed', self.__text_buffer_edited_cb)
> > +        self.textbuffer.set_text("TESTING")
> > +        self.textview.show()
> > +        self.pack_start(self.textview, expand=True, fill=True, padding=0)
> > +        self.show()
> > +
> > +    def __message_cb(self, collab, buddy, message):
> > +        action = message.get('action')
> > +	if action == 'entry_changed':
>
> Should I randomly generate the ID? Or is there a better way to assign ID's?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274/files#r47461956>
> .
>

erilyth replied:
Could you guide me as to how I can get the activity to assign the ids? Do you know of any other widget/activity which does this so I can take a look at it.

samdroid-apps replied:
Please make the activity pass the id to the init function.

On Mon, Dec 14, 2015, 8:06 PM Batchu Venkat Vishal <notifications@github.com>
wrote:

> In src/sugar3/activity/widgets.py
> <https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#discussion_r47473498>
> :
>
> > +        self._collab.connect('message', self.__message_cb)
> > +        self._collab.setup()
> > +        self.textview = Gtk.TextView()
> > +        self.textview.set_editable(True)
> > +        self.textview.set_cursor_visible(True)
> > +        self.textview.set_wrap_mode(Gtk.WrapMode.WORD)
> > +        self.textbuffer = self.textview.get_buffer()
> > +        self.textbuffer.connect('changed', self.__text_buffer_edited_cb)
> > +        self.textbuffer.set_text("TESTING")
> > +        self.textview.show()
> > +        self.pack_start(self.textview, expand=True, fill=True, padding=0)
> > +        self.show()
> > +
> > +    def __message_cb(self, collab, buddy, message):
> > +        action = message.get('action')
> > +	if action == 'entry_changed':
>
> Could you guide me as to how I can get the activity to assign the ids? Do
> you know of any other widget/activity which does this so I can take a look
> at it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274/files#r47473498>
> .
>

samdroid-apps commented on src/sugar3/activity/widgets.py:
```diff
@@ -396,3 +396,31 @@ def __init__(self):
+    def __message_cb(self, collab, buddy, message):
+        action = message.get('action')
+	if action == 'entry_changed':
+	    self.textbuffer.set_text(message.get('new_text'))
```
Do we preserve the users cursor position or selection?

If so, do we do that in a relative manner?  Eg.  If A types "foo bar", B selects "bar" then A delets the f of foo ("oo bar") - B should still have bar selected.

samdroid-apps replied:
Also, should we only send the changes as delta?  Is it fast enough now?

Also, think about race conditions:  eg. The text is "abcdefg".  A changes the a to 1 at exactly the same time as B changes the g to 7.  These change should not overwrite each other (at least they shouldn't most of the time).  Deltas will help solve this.

erilyth replied:
You mean I could send position as well as the change through the api and then appropriate changes could be performed on the other end?

erilyth replied:
My internet here is terrible so everything is slow, is it possible for you to test this once?

samdroid-apps commented on src/sugar3/activity/widgets.py:
```diff
@@ -396,3 +396,31 @@ def __init__(self):
+    def __text_buffer_edited_cb(self,textbuffer):
+	start_iter = textbuffer.get_start_iter()
+	end_iter = textbuffer.get_end_iter()
+	print textbuffer.get_text(start_iter, end_iter, True)
```
Please use logging.debug and add an appropriate explanatory prefix.

erilyth replied:
Where do these logs get stored? They are not visible in the Log activity right?

samdroid-apps replied:
On Sun, Dec 13, 2015, 10:38 PM Batchu Venkat Vishal <
notifications@github.com> wrote:

In src
<https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#discussion_r47443310>
/
<https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#discussion_r47443310>
sugar3
<https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#discussion_r47443310>
/activity/
<https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#discussion_r47443310>
widgets.py
<https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#discussion_r47443310>
:

> + self.textbuffer = self.textview.get_buffer() > +
self.textbuffer.connect('changed', self.__text_buffer_edited_cb) > +
self.textbuffer.set_text("TESTING") > + self.textview.show() > +
self.pack_start(self.textview, expand=True, fill=True, padding=0) > +
self.show() > + > + def __message_cb(self, collab, buddy, message): > +
action = message.get('action') > + if action == 'entry_changed': > +
self.textbuffer.set_text(message.get('new_text')) > + > + def
__text_buffer_edited_cb(self,textbuffer): > + start_iter =
textbuffer.get_start_iter() > + end_iter = textbuffer.get_end_iter() > +
print textbuffer.get_text(start_iter, end_iter, True)

Where do these logs get stored? They are not visible in the Log activity
right?

If you enable the debug log level, they will be stored.  See the
dev-environment docs if you want see how to do that.

—
Reply to this email directly or view it on GitHub
<https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274/files#r47443310>.<img
src="
https://ci4.googleusercontent.com/proxy/igndzllsiMlbQoNsnja6lSwqc118-lIkRYMzgIB96J27JaHMF8XTqNsje_Q00QNboPxkgBXKGndaV7BTIOJqzUxu4vYY9KE6_qX8-8H2ReDboQ3rIL_DJksykS1H3er6TWA194CHRYG0_-_YP2u3T_Q0_VmBbA=s0-d-e1-ft#https://github.com/notifications/beacon/AFvjmlxjI3bwPTNw2Be0saS9Nrpk_SSaks5pPVBCgaJpZM4G0RrG.gif
">

samdroid-apps commented:
Please consider making this into its own file in sugar3.graphics.

Please remember to write docs.

Thank you for getting started on a large feature!

I'll add more time to your taks, please feel free to make this a working branch and keep everybody updated on your progress.

samdroid-apps commented:
Wait, are you doing this through gci or just for yolo?  That's cool any way!

erilyth commented:
Also I wanted to know, the wrapper takes care of only allowing buddies who have joined this activity to edit the text right? I mean all that the widget needs to take care of is those post and receive data activities right?

erilyth commented:
I have made changes so now deletion and addition of texts are sent as deltas, and appropriate changes are made on the other end. Im unable to test this properly, if anyone could do that for me it would be great.

samdroid-apps commented:
On Mon, Dec 14, 2015, 12:53 AM Batchu Venkat Vishal <
notifications@github.com> wrote:

Also I wanted to know, the wrapper takes care of only allowing buddies who
have joined this activity to edit the text right? I mean all that the
widget needs to take care of is those post and receive data activities
right?

Only buddies that have joined that activity can edit the text.

The issue is that if A and B have been writing something together and then
C joins, C will only see an empty text box.  Maybe c should broadcast
action:init_requested and then it will take the 1st reply of
action:init_response?

—
Reply to this email directly or view it on GitHub
<https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#issuecomment-164261364>.<img
src="
https://ci6.googleusercontent.com/proxy/Tb2n447mL7JZJkv1QUPtjUkFWLv_m0SLWLVg_To0oIZy-S7b4MIqtBRoyKs5aWtpVFUFbii4AwPG6qCgkGP2VZsAIX8qRRvwUy-g4di3Xna7EEhspK1jq4UrfFdO_rsan9X9tYDKsIaH_XVty0Ya7SewABEqhw=s0-d-e1-ft#https://github.com/notifications/beacon/AFvjmpdXOBsgODS3zMl3Zx7yKtziMJHRks5pPW_vgaJpZM4G0RrG.gif
">

samdroid-apps commented:
By the way, how have you tested this code?  Have you got a demo activity I can use?

Great work so far.

Another good api would be set/get text.  However, the set text would have to affect all buddies.

erilyth commented:
Currently Im testing this by modifying the image viewer activity by adding the widget to it. Ill create a demo activity that you can use to test it.

samdroid-apps commented:
You may be interested in hello world activity on git.sugarlabs.org

On Mon, Dec 14, 2015, 4:06 PM Batchu Venkat Vishal <notifications@github.com>
wrote:

> Currently Im testing this by modifying the image viewer activity by adding
> the widget to it. Ill create a demo activity that you can use to test it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#issuecomment-164343499>
> .
>

erilyth commented:
Ok, Ill add this to the hello world activity and test it, if you want to test it as well just add 
"texteditor = CollabTextEditor(self)
self.set_canvas(texteditor)"
And import from "sugar3.graphics.texteditor import CollabTextEditor"

erilyth commented:
Now, when a buddy joins he will be sent the current contents of the editor and his instance will store the same values. For the author of the activity, the initialize_status=0 but as soon as he types something, his initialize_status=1, so he is all set.
Would you suggest a better approach or is this fine?

erilyth commented:
@samdroid-apps, were you able to test this?

samdroid-apps commented:
No sorry, I'm afk for this week.  I am happy to comment and review.  Great
job so far though.

Could somebody else test?

On Mon, Dec 14, 2015, 6:03 PM Batchu Venkat Vishal <notifications@github.com>
wrote:

> @samdroid-apps <https://github.com/samdroid-apps>, were you able to test
> this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#issuecomment-164363105>
> .
>

erilyth commented:
Regarding the multiple text editors, suppose there are 2 text editors A and B.
A -> sends a post, then does the message_cb of B gets toggled as well or only message_cb of A gets toggled (In another buddy who is sharing the activity)

samdroid-apps commented:
Yes, they both get toggled.  You could deal with this by prefixing the
action with an Id or maybe adding a "textview_id" prop to the message.

On Mon, Dec 14, 2015, 7:23 PM Batchu Venkat Vishal <notifications@github.com>
wrote:

> Regarding the multiple text editors, suppose there are 2 text editors A
> and B.
> A -> sends a post, then does the message_cb of B gets toggled as well or
> only message_cb of A gets toggled (In another buddy who is sharing the
> activity)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#issuecomment-164375800>
> .
>

erilyth commented:
[hello-world.zip](https://github.com/sugarlabs/sugar-toolkit-gtk3/files/61087/hello-world.zip)

Here is a sample app that uses the text editor, if anyone is willing to test this.

i5o commented:
Can you run pep8 on the code?
Also I will suggest a new pull request [not rebase or force push] with more concrete commits.
Maybe all with only one commit?

quozl commented:
-1, a vote against a new pull request.  a rebase with force push can correctly change the commit into just one if that's what the maintainer wants.

erilyth commented:
Yes sure I will do a rebase once someone can confirm this is working, I will also run a pep8 check as i5o mentioned.

samdroid-apps commented:
There seems to be lots of interesting work in the groupthink library (that appears to be old and unmaintained) regarding collaborative text editing.  I really have no idea what that library is, but you can view it here maybe:  https://github.com/163gal/Edit/blob/master/groupthink/stringtree.py

Does it have a git?  One with proper license headers and stuff?

samdroid-apps commented:
Ok, I got very excited.  It appears that it just does the same thing and nothing special (see the gtk_tools.py file).  Eg. it attaches positions to the insert and delete events.  But there is lots of interesting code in there, only confusingly burried under confusing abstractions.

walterbender commented:
The groupthink model of sharing is ""joined at the hip". Everyone sees
everything. It is buried in the Pippy activity collaboration if you want to
see an example.

On Sat, Dec 19, 2015 at 6:01 PM, Sam <notifications@github.com> wrote:

> Ok, I got very excited. It appears that it just does the same thing and
> nothing special (see the gtk_tools.py file). Eg. it attaches positions to
> the insert and delete events. But there is lots of interesting code in
> there, only confusingly burried under confusing abstractions.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sugarlabs/sugar-toolkit-gtk3/pull/274#issuecomment-166032552>
> .
>



-- 
Walter Bender
Sugar Labs
http://www.sugarlabs.org

erilyth commented:
Ok, sure Ill take a look at them. Are there anymore features like toolbars etc else that need to be done for the basic version of the text editor?

samdroid-apps commented on src/sugar3/activity/widgets.py:
```diff
@@ -396,3 +395,4 @@ def __init__(self):
         self.paste = PasteButton()
         self.insert(self.paste, -1)
         self.paste.show()
+
```
Please do not stylistic include changes to this file, unless you have other changes.

samdroid-apps commented on src/sugar3/graphics/texteditor.py:
```diff
@@ -0,0 +1,130 @@
+# Copyright (C) 2007, Red Hat, Inc.
```
It is an interesting choice to grant copyright of your work to Red Hat?  Have you been keeping it secret from when you used to work there in 2007 :smile:

erilyth replied:
I have no idea how that came in lol, I am a student getting ready for gsoc :P

erilyth commented:
Ran a PEP8 check on the code and adjusted the copyright information

samdroid-apps commented:
Have you tested this?  Please remember to add it to the makefile:

diff --git a/src/sugar3/graphics/Makefile.am b/src/sugar3/graphics/Makefile.am
index 0621750..8b3ee2e 100644
--- a/src/sugar3/graphics/Makefile.am
+++ b/src/sugar3/graphics/Makefile.am
@@ -20,6 +20,7 @@ sugar_PYTHON =              \
        radiotoolbutton.py      \
        scrollingdetector.py    \
        style.py                \
+       texteditor.py           \
        toggletoolbutton.py     \
        toolbarbox.py           \
        toolbox.py              \

samdroid-apps commented:
```
diff --git a/src/sugar3/graphics/Makefile.am b/src/sugar3/graphics/Makefile.am
index 0621750..8b3ee2e 100644
--- a/src/sugar3/graphics/Makefile.am
+++ b/src/sugar3/graphics/Makefile.am
@@ -20,6 +20,7 @@ sugar_PYTHON =              \
        radiotoolbutton.py      \
        scrollingdetector.py    \
        style.py                \
+       texteditor.py           \
        toggletoolbutton.py     \
        toolbarbox.py           \
        toolbox.py              \
```

samdroid-apps commented:
Have you tested this?  Typing on one computer does not sync with another, instead just generates errors.  Running this on fedora 24 rawhide with the activity zip provided above:

```
Traceback (most recent call last):
  File "/home/saam/sugar-build/build/out/install/lib/python2.7/site-packages/sugar3/graphics/texteditor.py", line 121, in __text_buffer_inserted_cb
    res_id=self._id, start_iter=start, new_text=text))
  File "/home/saam/sugar-build/build/out/install/lib/python2.7/site-packages/sugar3/presence/wrapper.py", line 243, in post
    self._text_channel.post(msg)
  File "/home/saam/sugar-build/build/out/install/lib/python2.7/site-packages/sugar3/presence/wrapper.py", line 271, in post
    self._send(json.dumps(msg))
  File "/usr/lib64/python2.7/json/__init__.py", line 243, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib64/python2.7/json/encoder.py", line 207, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib64/python2.7/json/encoder.py", line 270, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib64/python2.7/json/encoder.py", line 184, in default
    raise TypeError(repr(o) + " is not JSON serializable")
TypeError: <Gtk.TextIter object at 0x7ff1265ce6d0 (GtkTextIter at 0x7ffe1b828e20)> is not JSON serializable
```

erilyth commented:
Is there any possible way to send Gtk.TextIter objects via the collab api? Since it says its not JSON serializable. Or any conversion from these iter objects to text based objects?

erilyth commented:
@samdroid-apps , Do the set_data and get_data get automatically called when required or is there anything else that needs to be setup to call them explicitly when a buddy joins?

Currently it gives this error, even though I have implemented the set_data and get_data functions.
![screenshot from 2015-12-26 16-45-07](https://cloud.githubusercontent.com/assets/9864149/12006681/a416f61e-ac06-11e5-8b1c-44bf53a174ff.png)

erilyth commented on src/sugar3/graphics/texteditor.py:
```diff
@@ -0,0 +1,181 @@
+    '''
+    The leader of the activity sends the latest content to another buddy who joins.
+    '''
+    def get_data(self):
```
This get_data and set_data functions must be present in the activity itself and not the widget, but if I put them in the activity they are not connected to collab so errors occur, what should I do?

erilyth commented on src/sugar3/graphics/texteditor.py:
```diff
@@ -0,0 +1,167 @@
+            return
+        if self.has_initialized == False:
+            self.has_initialized = True
+        time.sleep(0.5)
```
Without this timeout, the joined buddy's collab instance is not being initialized completely, so it is not receiving the post signals in __message_cb. Should this be there?

samdroid-apps replied:
Hum, that is bad.  Maybe bind to the joined signal, which is called with somebody joins the activity.  You could then send a 'init_request' message and the reply could be send in __message_cb

erilyth replied:
I tried doing that but the joined signal was not being called at all

erilyth commented:
@samdroid-apps , The editor now works properly, during setup it takes about 1-2 seconds to get a copy of the activity owner's text editor instance and then updates are almost spontanious.

samdroid-apps commented:
This is really nice.  It works well and syncs together quickly.  Here are a few of my comments:

1.  You make a new collabwrapper instance for the activity for every text editor.  I think it would be better if the activity just passed their collab wrapper instance, so we don't end up creating lots of them.  I also believe that I make assumptions that there will only be 1 collab wrapper instance, so this could cause some werid things.
2.  Please make the CollabTextEditor inherit from Gtk.TextView, rather than having the self.textview thing
3.  It would be great if you provided a `set_text_synced` (or a better named alternative) function to set the text and sync that across all clients
4.  Could you make a standalone version?  See https://github.com/samdroid-apps/collabwrapper.  Making a standalone version would allow activities to use the wrapper right now.  You basically just have to replace `from sugar3.presence.wrapper import CollabWrapper` with `from collabwrapper import CollabWrapper`

I think that from the mailing list discussions about collabwrapper, that probably will miss this release (@tchx84?).  But this can probs be included in some activities standalone (browse for the descriptions?) and get shipped next release?

erilyth commented:
I made the editor inherit from the text view, added the specified function and made it standalone with try and except.
I tried making a single collab wrapper and passed it on to the widget, but a few problems arise, inserts are getting called twice, I will try debug this and update the PR.

samdroid-apps commented:
Looking great.

By making it standalone though, I did not mean add a standalone copy of the collabwrapper into sugar-toolkit-gtk3.  Could you please send a pull request with your collab text editor (as a new python file) to https://github.com/samdroid-apps/collabwrapper .  Then an activity will be able to include those 2 python scripts and not have to run with the patched sugar-toolkit-gtk3 version.

erilyth commented:
I added a PR to https://github.com/samdroid-apps/collabwrapper

Updated app to test the editor:
[hello-world.zip](https://github.com/sugarlabs/sugar-toolkit-gtk3/files/72601/hello-world.zip)

erilyth commented:
Updated app to test the editor: (Passes the collab instance to the editor)
[hello-world.zip](https://github.com/sugarlabs/sugar-toolkit-gtk3/files/72606/hello-world.zip)

samdroid-apps commented:
Development moved to https://github.com/samdroid-apps/collabwrapper based on the discourse in the "Current status of collaboration" thread
